internal/authentication/tokengetter.go

```diff
@@ -0,0 +1,92 @@
+package authentication
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	authenticationv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/utils/ptr"
+)
+
+type TokenGetter struct {
+	client            corev1.ServiceAccountsGetter
+	expirationSeconds int64
+	rotationThreshold time.Duration
+	tokens            map[types.NamespacedName]*authenticationv1.TokenRequestStatus
+	mu                sync.RWMutex
+}
+
+type TokenGetterOption func(*TokenGetter)
+
+// Returns a token getter that can fetch tokens given a service account.
+// The token getter also caches tokens which helps reduce the number of requests to the API Server.
+// In case a cached token is expiring a fresh token is created.
+func NewTokenGetter(client corev1.ServiceAccountsGetter, options ...TokenGetterOption) *TokenGetter {
+	tokenGetter := &TokenGetter{
+		client:            client,
+		expirationSeconds: int64(5 * time.Minute), // default token ttl
+		rotationThreshold: 1 * time.Minute,        // default rotation threshold
+		tokens:            map[types.NamespacedName]*authenticationv1.TokenRequestStatus{},
+	}
+
+	for _, opt := range options {
+		opt(tokenGetter)
+	}
+
+	return tokenGetter
+}
+
+func WithExpirationSeconds(expirationSeconds int64) TokenGetterOption {
+	return func(tg *TokenGetter) {
+		tg.expirationSeconds = expirationSeconds
+	}
+}
+
+func WithRotationThreshold(rotationThreshold time.Duration) TokenGetterOption {
+	return func(tg *TokenGetter) {
+		tg.rotationThreshold = rotationThreshold
+	}
+}
+
+// Get returns a token from the cache if available and not expiring, otherwise creates a new token
+func (t *TokenGetter) Get(ctx context.Context, key types.NamespacedName) (string, error) {
```

Comment:
Apologies for not noticing this earlier in the review, but Joe's earlier comment on the tokenLocks potentially growing unbounded made me realize that the tokens themselves could potentially grow unbounded with the current implementation since we only have a `Get` function.

What do you think about adding a method that could be used to "prune" expired tokens from the cache? 

I'm thinking this would be a method used as needed by a caller. I could imagine when we wire everything together we may want to ensure we prune any expired tokens that are older than X amount of time

Reply:
@joelanford What do you think?

Reply:
It appears that theres no way to invalidate tokens issued by the TokenRequest api 🤔 but for our purposes i guess we just remove keys i.e. namespaced names of the service accounts from the token getter cache

Reply:
Yeah, I'm thinking that it could simply be a process that runs every X amount of time and runs the method to ensure we aren't storing expired tokens for an unreasonable amount of time and adding bloat to the memory usage

Reply:
To clarify, I'm only recommending we implement the method that will delete the keys and tokens from the map that are expired and meet a certain criteria defined by the caller like expired for more than 2 hours

Reply:
Thinking about this a bit more, we would only remove entries from the cache when extensions are removed, in case there are expired tokens they would belong to service accounts of currently installed extensions. Further more the expired tokens will be replaced by fresh ones whenever an attempt is made to use the tokens so in theory tokens should not accumulate 🤔

Reply:
Do we need the complexity of a background goroutine? What if we just write a `reapExpiredTokens` function and call it every time `Get` is called?

Reply:
That would make it simpler. Ive added a `reapExpiredTokens` This also eliminates the need for `Clean` and `TokenExists`